lightning/src/ln/channel.rs

```diff
@@ -810,6 +810,31 @@ impl<Signer: Sign> Channel<Signer> {
 		self.channel_transaction_parameters.opt_anchors.is_some()
 	}
 
+	fn get_initial_channel_type(config: &UserConfig) -> ChannelTypeFeatures {
+		// The default channel type (ie the first one we try) depends on whether the channel is
+		// public - if it is, we just go with `only_static_remotekey` as its the only option
+		// available. If its private, we first try `scid_alias` as it provides better privacy with
+		// no other changes, and fall back to `only_static_remotekey`
+		let mut ret = ChannelTypeFeatures::only_static_remote_key();
+		if !config.channel_options.announced_channel && config.own_channel_config.negotiate_scid_alias {
+			ret.set_scid_alias_required();
+		}
+		ret
+	}
+
+	/// If we receive an error message, it may only be a rejection of the channel type we tried,
+	/// not of our ability to open any channel at all. Thus, on error, we should first call this
+	/// and see if we get a new `OpenChannel` message, otherwise the channel is failed.
+	pub(crate) fn advance_channel_type_pref(&mut self, chain_hash: BlockHash) -> Result<msgs::OpenChannel, ()> {
+		if !self.is_outbound() || self.channel_state != ChannelState::OurInitSent as u32 { return Err(()); }
```

Comment:
Hmmmm, I really prefer to keep as much channel state machine logic in `channel.rs` as possible. In general we've failed at this increasingly, but if I ever find time I'm gonna try to push a chunk back down. channelmanager should just be for inter-channel stuff, never have any real knowledge of channel's state machine transitions, though it has to sometimes ask about the current state.

Reply:
> because it assumes the error message is a response to open_channel. 

Heh, that's why i put the state check in `channel.rs` that way `channelmanager` doesn't really know what the error is in response to, it just tries and lets channel figure it out. I'm gonna rename `advance_channel_type_pref` to `maybe_handle_error_without_close` to make it "feel" more generic.

Unless you feel strongly I'm not gonna bother changing its return type, though, currently all of the `channel`/`channelmanager` bounday has chanellmanger "know" what type of message is being sent, cause its enforced by the type-checker. I agree in the future we should move some of those to events, but that's also a larger refactor.